sound/soc/sof/ipc4-topology.c

```diff
@@ -1065,6 +1065,51 @@ static int sof_ipc4_widget_assign_instance_id(struct snd_sof_dev *sdev,
 	return 0;
 }
 
+/* update hw_params based on the audio stream format */
+static int sof_ipc4_update_hw_params(struct snd_sof_dev *sdev, struct snd_pcm_hw_params *params,
+				     struct sof_ipc4_audio_format *fmt)
+{
+	struct snd_interval *i;
+	struct snd_mask *m;
+	snd_pcm_format_t snd_fmt;
+	unsigned int channels, rate;
+	int valid_bits;
+
+	/* update rate */
+	rate = fmt->sampling_frequency;
+	i = hw_param_interval(params, SNDRV_PCM_HW_PARAM_RATE);
+	i->min = rate;
+	i->max = rate;
+
+	/* update channel */
+	channels = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(fmt->fmt_cfg);
+	i = hw_param_interval(params, SNDRV_PCM_HW_PARAM_CHANNELS);
+	i->min = channels;
+	i->max = channels;
+
+	/* update fmt */
+	valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
```

Comment:
how on earth does this work? there is an unused 'valid bit' definition, and we use this super-complicated macro instead.

This is beyond confusing.